src/platform/w3dengine/client/w3ddisplay.cpp

```diff
@@ -60,80 +103,458 @@ W3DDisplay::W3DDisplay() :
-#ifdef GAME_DLL
-    Call_Method<void, W3DDisplay>(0x0073E360, this);
+    // TODO figure out gotos
+
+#ifdef BUILD_WITH_D3D8
+    if (g_applicationHWnd && IsIconic(g_applicationHWnd)) {
+        return;
+    }
+
+    static int lastFrame = -1;
+    static int syncTime = 0;
+
+    Update_Average_FPS();
+
+    if (g_theWriteableGlobalData->m_dynamicLOD && g_theGameLogic->Get_Dynamic_LOD()) {
+        g_theGameLODManager->Set_Dynamic_LOD_Level(g_theGameLODManager->Find_Dynamic_LOD_Level(m_averageFps));
+    } else {
+        g_theGameLODManager->Set_Dynamic_LOD_Level(DYNLOD_VERYHIGH);
+    }
+
+    if (g_theWriteableGlobalData->m_terrainLOD == TERRAIN_LOD_AUTOMATIC) {
+        if (g_theTerrainRenderObject) {
+            Calculate_Terrain_LOD();
+        }
+    }
+
+#ifdef GAME_DEBUG_STRUCTS
+    if (g_theWriteableGlobalData->m_doStatDump) {
+        // PerfStats.Log(false, true);
+        g_theWriteableGlobalData->m_doStatDump = false;
+    } else if (g_theWriteableGlobalData->m_doStats && g_theGameLogic->Get_Game_Mode() <= GAME_REPLAY) {
+        if (g_theGameLogic->Get_Frame()) {
+            if (!(g_theGameLogic->Get_Frame() % g_theWriteableGlobalData->m_statsInterval)) {
+                // PerfStats.Log(true, true);
+                g_theInGameUI->Message(L"-stats is running, at interval: %d.", g_theWriteableGlobalData->m_statsInterval);
+            }
+        }
+    }
+
+    if (m_debugDisplayCallback == Stat_Debug_Display || g_theWriteableGlobalData->m_benchmarkTimer > 0) {
+        Gather_Debug_Stats();
+    }
+#endif
+
+    bool time_frozen = g_theTacticalView->Is_Time_Frozen() && !g_theTacticalView->Is_Camera_Movement_Finished();
+    time_frozen |= g_theScriptEngine->Is_Time_Frozen_Debug() || g_theScriptEngine->Is_Time_Frozen_Script();
+    time_frozen |= g_theGameLogic->Is_Game_Paused();
+    time_frozen |= lastFrame == g_theGameClient->Get_Frame();
+    lastFrame = g_theGameClient->Get_Frame();
+
+    W3DView *primary_view = static_cast<W3DView *>(Get_First_View());
+
+    if (time_frozen && g_theScriptEngine->Is_Time_Fast()) {
+        primary_view->Update_Camera_Movements();
+        syncTime += g_theW3DFrameLengthInMsec;
+        return;
+    }
+
+    // Debug_Statistics::Begin_Statistics();
+
+    if (!g_theWriteableGlobalData->m_unkBool17) {
+        if (g_theTerrainTracksRenderObjClassSystem) {
+            g_theTerrainTracksRenderObjClassSystem->Update();
+        }
+
+        if (g_theTerrainRenderObject) {
+            if (g_theTerrainRenderObject->Get_Map()) {
+                if (g_theTerrainRenderObject->Get_Shroud()) {
+                    g_theTerrainRenderObject->Get_Shroud()->Render(primary_view->Get_3D_Camera());
+                }
+            }
+        }
+    }
+
+    if (!time_frozen) {
+        syncTime += g_theW3DFrameLengthInMsec;
+    }
+
+    W3D::Sync(syncTime);
+    int time = 30;
+    static unsigned int prevTime = rts::Get_Time();
+    unsigned int now;
+    now = rts::Get_Time();
```

Comment:
move to L388 so that we initialise directly.

Reply:
fixed.